File: examples/objc/BGExamples/LoginExtent.m
//
//  LoginExtent.m
//  BGExamples
//
//  Created by Sean Levin on 3/18/20.
//  Copyright © 2020 Verizon Media. All rights reserved.
//

#import "LoginExtent.h"

@implementation LoginExtent

/*
- (instancetype)init {
    self = [super init];
    if (self) {


    }
    return self;
}
*/
    // tag::login_enable_init[]
- (instancetype)initWithGraph:(BGGraph *)graph {
    self = [super initWithGraph:graph];
    
    _email = [self stateWithValue:@""];
    _password = [self stateWithValue:@""];
    
    // tag::login_enable_behavior[]
    [self behaviorWithDemands:@[self.email, self.password]
                     supplies:nil
                     runBlock:^(LoginExtent * _Nonnull extent) {
        // end::login_enable_init[]

        NSString *email = extent.email.value;
        NSString *password = extent.password.value;
        BOOL hasPassword = password.length > 0;
        BOOL loginEnabled = [extent validEmailAddress:email] && hasPassword;
        [extent sideEffect:@"enable login button" runBlock:^(LoginExtent * _Nonnull extent) {
            extent.login.loginButton.enabled = loginEnabled;
        }];
        
    }];
    // end::login_enable_behavior[]

    return self;
}

- (void)initBehaviorGraphComplete {
    
    _loggingIn = [self stateWithValue:@NO];
    
    // tag::login_complete_email[]
    _emailValid = [self stateWithValue:@NO];
    [self behaviorWithDemands:@[self.email]
                     supplies:@[self.emailValid]
                     runBlock:^(LoginExtent * _Nonnull extent) {
        
        NSString *email = extent.email.value;
        BOOL emailValid = [self validEmailAddress:email];
        [extent.emailValid updateValue:@(emailValid)];

    }];
    // end::login_complete_email[]

    _passwordValid = [self stateWithValue:@NO];
    [self behaviorWithDemands:@[self.password]
                     supplies:@[self.passwordValid]
                     runBlock:^(LoginExtent * _Nonnull extent) {
       
        NSString *password = extent.password.value;
        BOOL passwordValid = password.length > 0;
        [extent.passwordValid updateValue:@(passwordValid)];
        
    }];


    // tag::login_complete_enable[]
    _loginEnabled = [self stateWithValue:@NO];
    [self behaviorWithDemands:@[self.emailValid, self.passwordValid, self.loggingIn]
                     supplies:@[self.loginEnabled]
                     runBlock:^(LoginExtent * _Nonnull extent) {
        
        BOOL enabled = (extent.emailValid.value.boolValue &&
                        extent.passwordValid.value.boolValue &&
                        !extent.loggingIn.value.boolValue);
        [extent.loginEnabled updateValue:@(enabled)];
        [extent sideEffect:@"enable login button" runBlock:^(LoginExtent * _Nonnull extent) {
            extent.login.loginButton.enabled = extent.loginEnabled.value.boolValue;
        }];
        
    }];
    // end::login_complete_enable[]

    // tag::login_complete_login[]
    _loginClick = [self moment];
    _returnKey = [self moment];
    _loginComplete = [self moment];
    [self behaviorWithDemands:@[self.loginClick, self.returnKey, self.loginComplete]
                     supplies:@[self.loggingIn]
                     runBlock:^(LoginExtent * _Nonnull extent) {
        
        if ((extent.loginClick.justUpdated || extent.returnKey.justUpdated) &&
            extent.loginEnabled.traceValue.boolValue) {
            // Start login
            [extent.loggingIn updateValue:@YES];
        } else if (extent.loginComplete.justUpdated &&
                   extent.loggingIn.value.boolValue) {
            // Complete login
            [extent.loggingIn updateValue:@NO];
        }

        if ([extent.loggingIn justUpdatedTo:@YES]) {
            [extent sideEffect:@"login api call" runBlock:^(LoginExtent * _Nonnull extent) {
                [extent login:extent.email.value password:extent.password.value complete:^(BOOL success) {
                    [extent action:@"login complete" runBlock:^{
                        [extent.loginComplete updateValue:@(success)];
                    }];
                }];
            }];
        }

    }];
    // end::login_complete_login[]

    /*
     // Dont delete; its used in the documentation
     // this has an example of requireSync
     
    [self behaviorWithDemands:@[self.loginClick, self.returnKey, self.loginComplete]
                     supplies:@[self.loggingIn]
                     runBlock:^(LoginExtent t* _Nonnull extent) {
        
        if ((extent.loginClick.justHappened || extent.returnKey.justHappened) &&
            extent.loginEnabled.traceValue.boolValue) {
            // Start login
            [extent.loggingIn updateValue:@YES];
        } else if (extent.loginComplete.justHappened &&
                   extent.loggingIn.value.boolValue) {
            // Complete login
            [extent.loggingIn updateValue:@NO];
        }

        if ([extent.loggingIn justChangedTo:@YES]) {
            // tag::login_complete_loginalt[]
            [extent sideEffect:@"login api call" block:^(LoginExtent * _Nonnull extent) {
                [extent login:extent.email.value password:extent.password.value complete:^(BOOL success) {
                    [extent.graph action:@"login complete" requireSync:NO runBlock:^{
                        [extent.loginComplete happen:@(success)];
                    }];
                }];
            }];
            // end::login_complete_loginalt[]
        }

    }];
     */
}

- (instancetype)initShortWithGraph:(BGGraph *)graph {
    // tag::login_intro_short1[]
    [self behaviorWithDemands:@[self.email, self.password]
                     supplies:@[self.loginEnabled]
                     runBlock:^(LoginExtent *extent) {
        
        BOOL emailValid = [extent validEmailAddress:extent.email.value];
        BOOL passwordValid = extent.password.value.length > 0;
        BOOL enabled = emailValid && passwordValid;
        [extent.loginEnabled updateValue:@(enabled)];
        
    }];
    // end::login_intro_short1[]

    // tag::login_intro_short2[]
    [self behaviorWithDemands:@[self.loginClick]
                     supplies:@[self.loggingIn]
                     runBlock:^(LoginExtent * _Nonnull extent) {
        if (extent.loginClick.justUpdated && !extent.loggingIn.value.boolValue) {
            [extent.loggingIn updateValue:@YES];
        }
    }];
    
    [self behaviorWithDemands:@[self.email, self.password, self.loggingIn]
                     supplies:@[self.loginEnabled]
                     runBlock:^(LoginExtent *extent) {
        
        BOOL emailValid = [extent validEmailAddress:extent.email.value];
        BOOL passwordValid = extent.password.value.length > 0;
        BOOL enabled = emailValid && passwordValid && !self.loggingIn.value.boolValue;
        [extent.loginEnabled updateValue:@(enabled)];
        
    }];
    // end::login_intro_short2[]


    // tag::login_intro_sideeffect[]
    [self behaviorWithDemands:@[self.email, self.password, self.loggingIn]
                     supplies:@[self.loginEnabled]
                     runBlock:^(LoginExtent *extent) {

        BOOL emailValid = [extent validEmailAddress:extent.email.value];
        BOOL passwordValid = extent.password.value.length > 0;
        BOOL enabled = emailValid && passwordValid && !self.loggingIn.value.boolValue;
        [extent.loginEnabled updateValue:@(enabled)];

        [extent sideEffect:@"enable login button" runBlock:^(LoginExtent * _Nonnull extent) {
            extent.loginButton.enabled = extent.loginEnabled.value.boolValue;
        }];
    }];
    // end::login_intro_sideeffect[]

    return self;
}

// tag::login_intro_action[]
- (void)loginButtonClicked:(id)sender {
    [self.graph action:@"loginButtonClicked" runBlock:^{
        [self.loginClick update];
    }];
}
// end::login_intro_action[]

- (BOOL)validEmailAddress:(NSString *)email {
    return email.length > 0 && [email containsString:@"@"];
}

- (void)login:(NSString *)email password:(NSString *)password complete:(void(^)(BOOL success))complete {
    // some login api
}

// tag::login_sequence_compare[]
- (BOOL)emailChangedSincePassword {
    return self.email.event.sequence > self.password.event.sequence;
}
// end::login_sequence_compare[]

// tag::login_timestamp[]
- (NSDate *)loginCompletedWhen {
    return self.loginComplete.event.timestamp;
}
// end::login_timestamp[]

@end

File: examples/typescript/BGExamples/LoginExtent.ts
import { Graph, InitialEvent } from './behaveg/behavegjs';
import { Behavior } from "./behaveg/behavior"
import { State, Moment } from "./behaveg/resource"
import { Extent } from "./behaveg/extent"

// tag::login_enable_extent[]
class LoginExtent extends Extent {
    email: State<string>;
    password: State<string>;
// end::login_enable_extent[]

    loggingIn: State<boolean>
    emailValid: State<boolean>
    passwordValid: State<boolean>
    loginEnabled: State<boolean>
    loginClick: Moment;
    returnKey: Moment;
    loginComplete: Moment<boolean>;

    // tag::login_enable_init[]
    constructor(graph: Graph) {
        super(graph);

        this.email = new State("", this);
        this.password = new State("", this);

    // tag::login_enable_behavior[]
        this.makeBehavior([this.email, this.password], [], (extent: LoginExtent) => {
    // end::login_enable_init[]
            const email = extent.email.value;
            const password = extent.password.value;
            const hasPassword = password.length > 0;
            const loginEnabled = extent.validEmailAddress(email) && hasPassword;
            extent.sideEffect('enable login button', (extent) => {
                extent.enableLoginButton(loginEnabled);
            });
        });
    // end::login_enable_behavior[]
    }

    constructorComplete(graph: Graph) {

        this.loggingIn = new State(false, this);
    
        // tag::login_complete_email[]
        this.emailValid = new State(false, this);
        this.makeBehavior([this.email], [this.emailValid], (extent: this) => {
            const email = extent.email.value;
            const emailValid = extent.validEmailAddress(email);
            extent.emailValid.update(emailValid, true);
        });
        // end::login_complete_email[]

        this.passwordValid = new State(false, this);
        this.makeBehavior([this.password], [this.passwordValid], (extent: this) => {
            const password = extent.password.value;
            const passwordValid = password.length > 0;
            extent.passwordValid.update(passwordValid, true);
        });
        
        // tag::login_complete_enable[]
        this.loginEnabled = new State(false, this);
        this.makeBehavior([this.emailValid, this.passwordValid, this.loggingIn], [this.loginEnabled], (extent: this) => {
            const enabled = extent.emailValid.value && extent.passwordValid.value && !extent.loggingIn.value;
            extent.loginEnabled.update(enabled, true);
            extent.sideEffect('enable login button', (extent: this) => {
                extent.enableLoginButton(extent.loginEnabled.value);
            });
        });
        // end::login_complete_enable[]
    
        // tag::login_complete_login[]
        this.loginClick = new Moment(this);
        this.returnKey = new Moment(this);
        this.loginComplete = new Moment(this);
        this.makeBehavior([this.loginClick, this.returnKey, this.loginComplete], [this.loggingIn], (extent: this) => {
            if ((extent.loginClick.justUpdated || extent.returnKey.justUpdated) &&
                extent.loginEnabled.traceValue) {
                // Start login
                extent.loggingIn.update(true, true);
            } else if (extent.loginComplete.justUpdated &&
                       !extent.loginComplete.value &&
                       extent.loggingIn.value) {
                // Login failed
                extent.loggingIn.update(false, true);
            }
    
            if (extent.loggingIn.justUpdatedTo(true)) {
                extent.sideEffect('login api call', (extent: this) => {
                    extent.doLogin(extent.email.value, extent.password.value, (success: boolean) => {
                        extent.action('login call returned', () => {
                            extent.loginComplete.update(success);
                        });
                    });
                });
            }
        });
        // end::login_complete_login[]

        // Dont delete; its used in the documentation
        // this has an example of requireSync
        /*
        this.makeBehavior([this.loginClick, this.returnKey, this.loginComplete], [this.loggingIn], (extent: this) => {
            if ((extent.loginClick.justUpdated || extent.returnKey.justUpdated) &&
                extent.loginEnabled.traced.value) {
                // Start login
                extent.loggingIn.update(true, true);
            } else if (extent.loginComplete.justUpdated &&
                    !extent.loginComplete.value &&
                    extent.loggingIn.value) {
                // Login failed
                extent.loggingIn.update(false, true);
            }

            if (extent.loggingIn.justUpdatedTo(true)) {
                // tag::login_complete_loginalt[]
                extent.sideEffect('login api call', (extent: this) => {
                    extent.doLogin(extent.email.value, extent.password.value, (success: boolean) => {
                        extent.actionAsync('login call returned', () => {
                            extent.loginComplete.update(success);
                        });
                    });
                });
                // end::login_complete_loginalt[]
            }
        });
        */
 

    }

    constructorShort(graph) {

        // tag::login_intro_short1[]
        makeBehavior([email, password], [loginEnabled], (extent) => {

            const emailValid = validEmailAddress(email.value);
            const passwordValid = password.value.length > 0;
            const enabled = emailValid && passwordValid;
            loginEnabled.update(enabled);

        });
        // end::login_intro_short1[]
        
        
        // tag::login_intro_short2[]
        makeBehavior([loginClick], [loggingIn], (extent) => {
            if (loginClick.justUpdated && !loggingIn.value) {
                loggingIn.update(true);
            }
        });

        makeBehavior([email, password, loggingIn], [loginEnabled], (extent) => {

            const emailValid = validEmailAddress(email.value);
            const passwordValid = password.value.length > 0;
            const enabled = emailValid && passwordValid & !loggingIn.value;
            loginEnabled.update(enabled);

        });
        // end::login_intro_short2[]

        // tag::login_intro_action[]
        loginButton.onClick = () => {
            action("login button clicked", () => {
                loginClick.update();
            });
        };
        // end::login_intro_action[]

        // tag::login_intro_sideeffect[]
        makeBehavior([email, password, loggingIn], [loginEnabled], (extent) => {

            const emailValid = validEmailAddress(email.value);
            const passwordValid = password.value.length > 0;
            const enabled = emailValid && passwordValid & !loggingIn.value;
            loginEnabled.update(enabled);
            
            extent.sideEffect("login button enabled", (extent) => {
                loginButton.enabled = loginEnabled.value;
            });

        });
        // end::login_intro_sideeffect[]
    
    }

    constructorCompleteShort(graph: Graph) {

        // tag::login_complete_short[]
        this.email = new State(this, "");
        this.password = new State(this, "");
        this.loggingIn = new State(this, false);
    
        this.loginEnabled = new State(this, false);
        this.makeBehavior([this.email, this.password, this.loggingIn], 
                          [this.loginEnabled], 
                          (extent: this) => {
            const emailValid = extent.validEmailAddress(this.email.value);
            const passwordValid = extent.password.value.length > 0;
            const enabled = emailValid && passwordValid && !extent.loggingIn.value;
            extent.loginEnabled.update(enabled, true);
            extent.sideEffect('enable login button', (extent: this) => {
                extent.enableLoginButton(extent.loginEnabled.value);
            });
        });
    
        this.loginClick = new Moment(this);
        this.loginComplete = new Moment(this);
        this.makeBehavior([this.loginClick, this.loginComplete], 
                          [this.loggingIn], 
                          (extent: this) => {
            if (extent.loginClick.justUpdated &&
                extent.loginEnabled.traceValue) {
                // Start login
                extent.loggingIn.update(true, true);
            } else if (extent.loginComplete.justUpdated &&
                       !extent.loginComplete.value &&
                       extent.loggingIn.value) {
                // Login failed
                extent.loggingIn.update(false, true);
            }
    
            if (extent.loggingIn.justUpdatedTo(true)) {
                extent.sideEffect('login api call', (extent: this) => {
                    extent.doLogin(extent.email.value, extent.password.value, (success: boolean) => {
                        extent.actionAsync('login call returned', () => {
                            extent.loginComplete.update(success);
                        });
                    });
                });
            }
        });
        // end::login_complete_short[]
    }

    validEmailAddress(email: string) : boolean {
        return email.length > 0 && email.includes('@');
    }

    doLogin(email: string, password: string, complete: (boolean) => void) {
       // login api calls
    };

    enableLoginButton(enabled: boolean) {
        // side effect to set the enabled state of the login button
    }

    // tag::login_sequence_compare[]
    emailChangedSincePassword() : boolean {
        return this.email.event.sequence > this.password.event.sequence;
    }
    // end::login_sequence_compare[]
    
    // tag::login_timestamp[]
    loginCompletedWhen() : Date {
        return this.loginComplete.event.timestamp;
    }
    // end::login_timestamp[]

}

class LoginPage {

    graph: Graph;
    loginExtent: LoginExtent;

    constructor() {
    // tag::login_enable_setup[]
        this.graph = new Graph();
        this.loginExtent = new LoginExtent(this.graph);
        this.graph.action('new login page', () => {
            this.loginExtent.addToGraph();
        });
    // end::login_enable_setup[]
    }
    
    // tag::login_enable_actions[]
    didUpdateEmailField(contents: string) {
        this.graph.action('update email field', () => {
            this.loginExtent.email.update(contents, true);
        });
    }

    didUpdatePasswordField(contents: string) {
        this.graph.action('update password field', () => {
            this.loginExtent.password.update(contents, true);
        });
    }
    // end::login_enable_actions[]
    
    // tag::login_complete_click[]
    loginButtonClicked() {
        this.graph.action('login button clicked', () => {
            this.loginExtent.loginClick.update();
        });
    }
    // end::login_complete_click[]
    

}
